### ISSUE TITLE
After the Built-in PF (Packet Filter) is removed, is there any alternative?

### ISSUE THREAD

**Issue by KaiveYoung** [OpenVPN/openvpn#314]
I have been using PF to implement permission isolation between clients. Generate pf rules through the database. After upgrading to 2.6, PF has been removed. Is there any alternative?

**Comment by schwabe** [comment]
It had been removed since we sent several heads up to the mailing list if anyone is using it and nobody answered it and generally operating systems nowadays offer similar (or much better) functionality in filtering IP packets.

**Comment by KaiveYoung** [comment]
@cron2 @schwabe There are thousands of edge devices and dozens of administrator accounts in my vpn network. Administrators and devices are distinguished by groups, and administrators are forbidden to access each other. Edge devices are forbidden to access each other. Different groups are also forbidden to access each other. The important thing is that the group to which it belongs changes dynamically. It may not be possible to complete such a complicated function by distinguishing by IP Packet.

**Comment by schwabe** [comment]
The OpenVPN packet was also just an IP filter. Everything that could be accomplished by OpenVPN's own packet filter can also be achieved by an operating system IP filter. What you are describing are quite typical rules people implement in packet filters. if you really think that OpenVPN packet filter was supirior, please give us more details.

**Comment by KaiveYoung** [comment]
@schwabe 
When the IP changes dynamically or the number of devices continues to increase, and the device permission group cannot be distinguished by CIDR, the rules of iptables will be very complicated

**Comment by schwabe** [comment]
You are still not explaining how was the OpenVPN PF filter was better for you there.

**Comment by dsommers** [comment]
@KaiveYoung it is possible to achieve what you want using the scripts interfaces, like `--learn-address` script hook and run a script which updates the firewall rules dynamically.

**Comment by MrGrosNounours** [comment]
Just like KaiveYoung, I've chosen to adopt this solution to filter traffic among my numerous users, exceeding the 1000 mark. With over a decade of experience in its usage, OpenVPN's built-in firewall has proven to be the most suitable method for maintaining security and managing constantly changing IP addresses in my network, thereby avoiding potential complexities associated with other approaches like iptables.

**Comment by schwabe** [comment]
@MrGrosNounours that is very unfortunate now but we made several attempts to find anyone that was still using the feature, so we deprecated it. I think it will only come back if someone invests time in it to bring the feature back.

**Comment by igor-kremin** [comment]
I also use a packet filter to restrict access to certain networks, this is the best solution I could implement. It's a shame that such useful functionality was removed. It was not perfect, but it was very useful and did its job well. The mailing list does not cover all users who use OpenVPN; this is a tool that have configured and it works; you return to it only when absolutely necessary. If there is an opportunity to return this wonderful functionality, we would love to get it back.

**Comment by cron2** [comment]
If you want it back, we need someone who is willing to maintain that code, add IPv6 to it, and be responsible for it.

Since none of the core developers have a particular interest in it (we use the firewalling component of the server OS, like Linux or FreeBSD to achieve the desired filtering effects), we can't maintain it.  Our time is limited.

**Comment by HESS-BEA** [comment]
I am also using this feature really hard.
Also rewrote some little code for using it in a different way :)

Now i can make "Rule Files" based on common name of the client, and allowing in it some different CIDRs based on his name.
My whole Server is splited in 18 Sub VPN Networks. Devices and Users.

Users can access Devices based on their Rule by Common Name and Devices are currently only allowed to access the Gateway and comminucate to Users.


In a nutshell, i rewrote the pk stuff like, that i am having a simple "rule management system" based on files (pf files) they created by a Script / Database.

**Comment by schwabe** [comment]
It would have been good if you contributed these changes back to the project. Since nobody worked on the feature or contributed anything back or speak up when we scheduled it for removal, it was deemed to not be used by anyone.

If the implementation is modernised and provides benefit, a reintroduction can be discussed if we have an active maintainer for this.

**Comment by HESS-BEA** [comment]
Sorry Schwabe, i only mad a little patch to 2.4 and 2.5 for using that Feature with the common name. Already some Months / Years ago.

I only remember that under certain circumstances somehow IP addresses were not used as PF filters, but only common names.

However, I wanted a combination of both.
Clients are dynamic via "DHCP" and common name, but should then only have access to certain VPN servers, which are separated by IP subnets.

I can't remember the circumstances of the patch, but maybe it will help:

[allow_subnets_in_c2c_25.patch](https://github.com/user-attachments/files/16012652/allow_subnets_in_c2c_25.patch)
[allow_subnets_in_c2c_24.patch](https://github.com/user-attachments/files/16012653/allow_subnets_in_c2c_24.patch)

**Comment by schwabe** [comment]
@HESS-BEA 2.4 and 2.5 are end of line by now. We will probably not look at these patches but it is good to have them as reference if someone else wants to revive the packet filtering in OpenVPN 2.x

**Comment by khvalera** [comment]
Too bad, it was a very useful feature :(
